Add alarmdecoder brand specific arming sequences

ajschmidt8 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Previously the `autobypass` configuration option only worked for `home` and `away` arming. It now works for `night` arming also. Review the `alarmdecoder` integration docs to make sure the `autobypass` option is configured to your preference.

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR replaces AlarmDecoder's `AlarmDecoder` class with the extended `AdExt` class from [adext](https://pypi.org/project/adext/). This enables more arming options for the AlarmDecoder integration. It also enables the `code_arm_required` option to work with DSC alarm panels. Previously, it only worked for Honeywell panels.

The new arming sequences are spelled out in the new docs (https://github.com/home-assistant/home-assistant.io/pull/13725).


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
alarmdecoder:
  device:
    type: socket
    host: <ip>
    port: 10000
  code_arm_required: false
  autobypass: true
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/issues/35728
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13725

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ajschmidt8 committed:
add adext

ajschmidt8 committed:
Merge remote-tracking branch 'upstream' into add-adext

MartinHjelmare commented on homeassistant/components/alarmdecoder/__init__.py:
```diff
@@ -110,6 +112,9 @@
                     CONF_CODE_ARM_REQUIRED, default=DEFAULT_CODE_ARM_REQUIRED
                 ): cv.boolean,
+                vol.Optional(
+                    CONF_ALT_NIGHT_MODE, default=DEFAULT_ALT_NIGHT_MODE
```
We no longer accept additions or changes to the yaml config of integrations. See: https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

This integration needs to be refactored to use a config flow and config entries.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

I suggest removing this option for now so we can go ahead with this PR. Then make a separate PR where config entries and config flow is introduced.

ajschmidt8 replied:
@MartinHjelmare any chance I can get an exception here? I introduced this functionality previously in https://github.com/home-assistant/core/pull/35761 and was told that it needed to be moved to a separate library, which I've now done. I'm eager to get this merged because I've heard from people who were unhappy about the changes I made in https://github.com/home-assistant/core/pull/32292 and this will give them a way to revert those changes. It's starting to seem like the bars are getting raised here every time I open a PR. This is also the second time I've had to open up my long documentation PR here https://github.com/home-assistant/home-assistant.io/pull/13725. I'm not looking forward to revising it now and opening it up a 3rd time for the same functionality.

My next task is to bring multiple partition support to AlarmDecoder which will require more config changes, so I'd be happy to introduce the config flow then. But I'm really hoping you can make an exception for this code.

MartinHjelmare replied:
Sorry, we're following our design guidelines and architecture decisions. When the project develops our standards evolve and changes to integrations need to follow the new standards.

ajschmidt8 committed:
Merge remote-tracking branch 'upstream/dev' into add-adext

ajschmidt8 committed:
rm new config parameter

ajschmidt8 commented:
@MartinHjelmare, I've removed the new config parameter for the time being. I've also converted the PR to a draft because I want to test this change locally to make sure everything is still working as expected. Are there any other changes that would need to be made before this could be merged?

MartinHjelmare commented:
Looks good now!

ajschmidt8 commented:
@MartinHjelmare, tested ready for official review.

ajschmidt8 commented:
sorry, testing one more quick thing. i was hoping to finish before you saw my comment.

ajschmidt8 committed:
update adext

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
Please extend the breaking change paragraph with what the user needs to do to cope with the breaking change.

ajschmidt8 commented:
> Please extend the breaking change paragraph with what the user needs to do to cope with the breaking change.

@MartinHjelmare, I'm not sure there's anything to do to cope with this change. It's mostly just a functionality change to be aware of. I reviewed the original `autobypass` PR #30002 and didn't see a reason why `autobypass` wasn't included in night mode originally. I added it to night mode in my new logic just so it's consistent across all arming modes. Let me know what you'd like me to do here.

I also have a question about the `requirements` field in `manifest.json`. I replaced `alarmdecoder` with `adext`, however there are still items imported from `alarmdecoder` in this integration (like [here](https://github.com/home-assistant/core/pull/36692/files#diff-0084eacb87a2f8212f5bd9e29ca3282bR6-R7)). `alarmdecoder` is [listed as a dependency](https://github.com/ajschmidt8/adext/blob/master/setup.py#L21) in my `adext` package, so I don't foresee this being an issue, but I wanted to double check.

MartinHjelmare commented:
How will users be affected by the breaking change?

ajschmidt8 commented:
@MartinHjelmare updated!

MartinHjelmare commented:
I would advice users to review the autobypass config option to make sure it's set according to their preference.

ajschmidt8 commented:
> I would advice users to review the autobypass config option to make sure it's set according to their preference.

@MartinHjelmare done!

can you check this 3-line PR also? #36695

ajschmidt8 commented:
@MartinHjelmare, according to [your comment from this thread](https://github.com/home-assistant/core/issues/32819#issuecomment-619388037), I don't believe it's possible to fully convert AlarmDecoder to Config Flow yet. I can't figure out a way to get the `zones` config option shown below to be configurable through Config Flow. Is there a way to do this currently? If not, can I get an exception to add the feature we discussed above into the existing yaml configuration for AlarmDecoder?


```yaml
# Example configuration.yaml entry
alarmdecoder:
  device:
    type: socket
    host: 192.168.1.20
    port: 10000
  panel_display: false
  zones:
    01:
      name: 'Smoke Detector'
      type: 'smoke'
      rfid: '0123456'
    02:
      name: 'Front Door'
      type: 'opening'
```
